capa/features/freeze/__init__.py

```diff
@@ -0,0 +1,388 @@
+        frozen = True
+
+
+class AddressType(str, Enum):
+    ABSOLUTE = "absolute"
+    RELATIVE = "relative"
+    FILE = "file"
+    DN_TOKEN = "dn token"
+    DN_TOKEN_OFFSET = "dn token offset"
+    NO_ADDRESS = "no address"
+
+
+class Address(HashableModel):
+    type: AddressType
+    value: Any
+
+    @classmethod
+    def from_capa(cls, a: capa.features.address.Address) -> "Address":
+        if isinstance(a, capa.features.address.AbsoluteVirtualAddress):
+            return cls(type=AddressType.ABSOLUTE, value=int(a))
+
+        elif isinstance(a, capa.features.address.RelativeVirtualAddress):
+            return cls(type=AddressType.RELATIVE, value=int(a))
+
+        elif isinstance(a, capa.features.address.FileOffsetAddress):
+            return cls(type=AddressType.FILE, value=int(a))
+
+        elif isinstance(a, capa.features.address.DNTokenAddress):
+            # TODO: probably need serialization here
+            return cls(type=AddressType.DN_TOKEN, value=a.token)
+
+        elif isinstance(a, capa.features.address.DNTokenOffsetAddress):
+            # TODO: probably need serialization here
+            return cls(type=AddressType.DN_TOKEN_OFFSET, value=(a.token, a.offset))
+
+        elif a == capa.features.address.NO_ADDRESS or isinstance(a, capa.features.address._NoAddress):
+            return cls(type=AddressType.NO_ADDRESS, value=None)
+
+        elif isinstance(a, capa.features.address.Address) and not issubclass(type(a), capa.features.address.Address):
+            raise ValueError("don't use an Address instance directly")
+
+        elif isinstance(a, capa.features.address.Address):
+            raise ValueError("don't use an Address instance directly")
+
+        else:
+            assert_never(a)
+
+    def to_capa(self) -> capa.features.address.Address:
+        if self.type is AddressType.ABSOLUTE:
+            return capa.features.address.AbsoluteVirtualAddress(self.value)
+
+        elif self.type is AddressType.RELATIVE:
+            return capa.features.address.RelativeVirtualAddress(self.value)
+
+        elif self.type is AddressType.FILE:
+            return capa.features.address.FileOffsetAddress(self.value)
+
+        elif self.type is AddressType.DN_TOKEN:
+            return capa.features.address.DNTokenAddress(self.value)
+
+        elif self.type is AddressType.DN_TOKEN_OFFSET:
+            return capa.features.address.DNTokenOffsetAddress(*self.value)
+
+        elif self.type is AddressType.NO_ADDRESS:
+            return capa.features.address.NO_ADDRESS
+
+        else:
+            assert_never(self.type)
+
+    def __lt__(self, other: "Address") -> bool:
+        if self.type != other.type:
+            return self.type < other.type
+
+        if self.type in (AddressType.ABSOLUTE, AddressType.RELATIVE, AddressType.FILE):
+            return self.value < other.value
+
+        elif self.type in (AddressType.DN_TOKEN, AddressType.DN_TOKEN_OFFSET):
+            return self.value < other.value
```

Comment:
```suggestion
        elif self.type is AddressType.DN_TOKEN:
            return self.value.value < other.value.value

        elif self.type is AddressType.DN_TOKEN_OFFSET:
            self_token, self_offset = self.value
            other_token, other_offset = other.value
            return  (self_token.value, self_offset) < (other_token.value, other_offset)
```
we need to update dncil to support `__int__`, `__eq__`, etc. to help clean this up. but for now we need to compare the token value and not the `Token` object.

Reply:
good point. also need to find a test case. i'll work on this.